docs/examples/reusable-components.md

```diff
@@ -0,0 +1,262 @@
+
+1. `fromString`: converts a string option to type `O`.
+2. `onChange0`: performs an action based on selected option and state.
+3. `options`: list of options.
+4. `selected0`: currently selected option function of the state.
+
+### Button
+
+Likewise for a button component inputs:
+
+1. `onClick0`: react to click events by performing an action.
+2. `isDisabled`: disable the button based on the state.
+
+```scala mdoc:js:shared
+import org.scalajs.dom
+import cats.Show
+import cats.kernel.Eq
+import cats.syntax.all._
+
+// Class defining all components
+class Components[F[_], S, A] {
+
+  // Custom select component
+  def customSelect[O: Show: Eq](
+      fromString: String => Option[O],
+      onChange0: (S, O) => Option[A],
+      options: List[O],
+      selected0: S => O
+  )(implicit dsl: ff4s.Dsl[F, S, A]): dsl.V = {
+    import dsl._
+    import dsl.html._
+    useState { state =>
+      select(
+        onChange := ((ev: dom.Event) =>
+          ev.target match {
+            case el: dom.HTMLSelectElement =>
+              fromString(el.value).flatMap(onChange0(state, _))
+            case _ => None
+          }
+        ),
+        options.map { name =>
+          option(
+            selected := (name === selected0(state)), // use of Eq[O] implicit
+            key := name.show, // use of Show[O] implicit
+            value := name.show,
+            name.show
+          )
+        }
+      )
+    }
+
+  }
+
+  // Custom button component
+  def customButton(
+      dsl: ff4s.Dsl[F, S, A]
+  )(child: dsl.V, onClick0: S => Option[A], isDisabled: S => Boolean): dsl.V = {
+    import dsl._
+    import dsl.html._
+
+    useState { state =>
+      button(
+        child,
+        disabled := isDisabled(state),
+        onClick := (_ => onClick0(state))
+      )
+    }
+  }
+
+}
+```
+
+Note that an instance of `ff4s.Dsl` is passed explicitly to the `customButton` function
+as the `child` argument type depends explicitly on it, contrary to the select component example where
+the instance is passed as an implicit.
+
+## State
+
+```scala mdoc:js:shared
+final case class State(
+    numOfRequests: Int = 0,
+    currencyPair: CurrencyPair = CurrencyPair.EURUSD,
+    apiResponse: Option[ApiResponse] = None
+) {
+
+  def ccyPairOption: (String, String) =
+    (
+      currencyPair.show.toString.take(3),
+      currencyPair.show.toString.takeRight(3)
+    )
+
+}
+```
+
+```scala mdoc:js:shared
+import io.circe._
+import io.circe.generic.semiauto._
+
+final case class ApiResponse(rates: Map[String, Double])
+object ApiResponse {
+  implicit val decoder: Decoder[ApiResponse] = deriveDecoder
+}
+```
+
+In this example, the elements of `customSelect` are currency pairs. A generic currency pair is modelled by a
+`CurrencyPair` trait. Specific pairs are `case object`'s extending from the latter.
+
+```scala mdoc:js:shared
+import cats.Show
+import cats.kernel.Eq
+
+sealed trait CurrencyPair
```

Comment:
I know this is a toy example, but I'm not very fond of this encoding. There are ~180 currencies, so this wouldn't scale in practice.